sys/senml/senml.c

```diff
@@ -0,0 +1,328 @@
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
+#include <ctype.h>
+#include <stdio.h>
+
+#include "senml.h"
+#include "jsmn.h"
+#include "cbor.h"
+
+#define ENABLE_DEBUG  (0)
+#include "debug.h"
+
+
+int senml_decode_json_s(char *input, senml_pack_t *pack)
+{
+    jsmn_parser parser;
+
+    jsmn_init(&parser);
+
+    int num_toks = jsmn_parse(&parser, input, strlen(input), NULL, 65535);
+
+    if (num_toks == JSMN_ERROR_INVAL || num_toks == JSMN_ERROR_PART) {
+        DEBUG("ERROR: unable to parse JSON string #1\n");
+        return -1;
+    }
+
+    jsmntok_t tokens[num_toks];
+
+    jsmn_init(&parser);
+
+    jsmn_parse(&parser, input, strlen(input), tokens, num_toks);
+
+    if (num_toks == JSMN_ERROR_INVAL || num_toks == JSMN_ERROR_PART) {
+        DEBUG("ERROR: unable to parse JSON string #2\n");
+        return -1;
+    }
+
+    if (tokens[0].type != JSMN_ARRAY) {
+        DEBUG("ERROR: not a valid SenML document\n");
+        DEBUG("type is: %d\n", (int)tokens[0].type);
+        return -1;
+    }
+
+    int array_num = -1;
+
+    jsmntype_t last_type = JSMN_UNDEFINED;
+
+    char *tmp_key, *tmp_val;
+
+    bool base_values_checked = false;
+
+    for (size_t i = 0, j = num_toks; j > 0; i++, j--) {
+        jsmntok_t *t = &tokens[i];
+
+        if (t->type == JSMN_ARRAY) {
+            last_type = JSMN_ARRAY;
+            continue;
+        }
+
+        if (t->type == JSMN_OBJECT) {
+            array_num++;
+            last_type = JSMN_OBJECT;
+            continue;
+        }
+
+        if (last_type == JSMN_OBJECT && t->type == JSMN_STRING) {
+            tmp_key = &input[t->start];
+            input[t->end] = '\0';
+            i++;
+            j--;
+            t = &tokens[i];
+            tmp_val = &input[t->start];
+            input[t->end] = '\0';
+
+            if (array_num == 0 && !base_values_checked) {
+                if (!strcmp(tmp_key, SJ_VERSION)) {
+                    pack->base_info->version = (uint8_t)atoi(tmp_val);
+                }
+                else if (!strcmp(tmp_key, SJ_BASE_NAME)) {
+                    pack->base_info->base_name = tmp_val;
+                }
+                else if (!strcmp(tmp_key, SJ_BASE_TIME)) {
+                    pack->base_info->base_time = atof(tmp_val);
+                }
+                else if (!strcmp(tmp_key, SJ_BASE_UNIT)) {
+                    pack->base_info->base_unit = tmp_val;
+                }
+                else if (!strcmp(tmp_key, SJ_BASE_VALUE)) {
+                    pack->base_info->base_value = atof(tmp_val);
+                }
+                else {   // no more base information
```

Comment:
Because [“An object is an unordered set of name/value pairs.”](http://www.json.org/), the follwoing should be valid:

```json
[
  { "t": 1276020076, "v":23.5, "u":"Cel", "bn": "urn:dev:ow:10e2073a01080063" }
  { "t": 1276020091, "v":23.6, "u":"Cel" }
]
```